pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the ``DecompositionRule`` class to represent a decomposition rule."""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import Callable
+
+from pennylane.operation import Operator
+from .resources import Resources, CompressedResourceOp, make_resource_rep
+
+
+def decomposition(qfunc: Callable, resource_fn: Callable = None) -> DecompositionRule:
+    """Creates a decomposition rule from a quantum function.
```

Comment:
@isaacdevlugt Can you read this over to see if it is easy to understand?

Reply:
Taking a look!